### PR TITLE
chore(exitCodes): adding exit code for browser connect errors

### PR DESCRIPTION
**Pull request by cnishina**

### CODE REVIEW

**Comment by cnishina** [comment]
Unit tests added for BrowserError for direct connect.

**Comment by cnishina** [comment]
depends on #3131

File: spec/unit/driverProviders/direct_test.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
+
+var path = require('path');
+var fs = require('fs');
+var BrowserError = require('../../../built/exitCodes').BrowserError;
+
+var removeSeleniumFolder = function() {
```
Comment (juliemr) [inline spec/unit/driverProviders/direct_test.js]: Instead of doing this, just use a config file that points at a different (and invalid) chromedriver location.

**Comment by cnishina** [comment]
Instead of using the `selenium` folder, creating a "chromedriver" file in the `os.tmpdir()` for my unit test.

File: spec/unit/driverProviders/direct_test.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        expect(e.code).toBe(BrowserError.CODE);
+      }
+      expect(errorFound).toBe(true);
+    });
+  });
+
+  describe('with selenium drivers', function() {
+    var chromedriver = '';
+    beforeEach(function() {
+      // add files to selenium folder
```
Comment (juliemr) [inline spec/unit/driverProviders/direct_test.js]: Why is it important that the file have the correct version? Aren't we just testing that if the file exists but isn't a proper driver, we get the right error code?

I think this should either just use a bogus file that we create (like os.tpdir() + '/foo/bar') or not specify `chromeDriver` in the config and let the driverProvider use its normal logic.

File: lib/driverProviders/browserStack.ts [hunk 3]
```diff
@@ -3,10 +3,11 @@
  * It is responsible for setting up the account object, tearing
  * it down, and setting up the driver correctly.
  */
-import * as request from 'request';
+import * as https from 'https';
```
Comment (juliemr) [inline lib/driverProviders/browserStack.ts]: Can this 'request' removal be split into its own PR?

File: lib/exitCodes.ts [hunk 4]
```diff
@@ -1,28 +1,79 @@
 import {Logger} from './logger2';
 
 const CONFIG_ERROR_CODE = 105;
+const BROWSER_ERROR_CODE = 135;
```
Comment (juliemr) [inline lib/exitCodes.ts]: I think `BROWSER_ERROR` is too ambiguous - let's make this `BROWSER_CONNECT_ERROR_CODE`. It's long but our users shouldn't need to type it.

File: lib/exitCodes.ts [hunk 5]
```diff
@@ -1,28 +1,79 @@
...
+  static DEFAULT_MSG = 'protractor encountered an error';
 
-export class ProtractorError {
   error: Error;
   description: string;
   code: number;
   stack: string;
   constructor(logger: Logger, description: string, code: number) {
+    super();
     this.error = new Error();
```
Comment (juliemr) [inline lib/exitCodes.ts]: We have both inheritance and composition - if we're inheriting from error can we just use `this.stack`?

Comment (juliemr) [inline lib/exitCodes.ts]: Or, not inherit?

File: lib/exitCodes.ts [hunk 6]
```diff
@@ -1,28 +1,79 @@
...
 /**
  * Configuration file error
  */
 export class ConfigError extends ProtractorError {
+  static DEFAULT_MSG = 'configuration error';
   static CODE = CONFIG_ERROR_CODE;
-  constructor(logger: Logger, description: string) {
-    super(logger, description, ConfigError.CODE);
+  constructor(logger: Logger, opt_msg?: string) {
+    super(logger, opt_msg || ConfigError.DEFAULT_MSG, ConfigError.CODE);
```
Comment (juliemr) [inline lib/exitCodes.ts]: Why have a default configuration msg? 'configuration error' is pretty useless by itself, I'd say keep this required.

File: lib/exitCodes.ts [hunk 7]
```diff
@@ -1,28 +1,79 @@
...
+    process.exit(ConfigError.CODE);
+  }
+}
+
+/**
+ * Browser errors including getting a driver session, direct connect, etc.
+ */
+export class BrowserError extends ProtractorError {
+  static DEFAULT_MSG = 'browser error';
```
Comment (juliemr) [inline lib/exitCodes.ts]: Ditto for not having a default error, make the error-thrower say something useful.

File: lib/launcher.ts [hunk 8]
```diff
@@ -176,7 +177,18 @@ let initFn = function(configFile: string, additionalConfig: Config) {
...
+          if (errorCode) {
+            logger.error(e.stack);
+            process.exit(errorCode);
+          } else {
+            logger.error(e.stack);
+          }
+        });
+
         process.on('exit', (code: number) => {
+          console.log(code);
```
Comment (juliemr) [inline lib/launcher.ts]: Should we be logging this? If someone wants to use it, they can just use the exit code.

Comment (juliemr) [inline lib/launcher.ts]: And it's noted below.

**Comment by juliemr** [comment]
A couple clean-up comments.

File: lib/driverProviders/browserStack.ts (hunk 3 above)
Comment (cnishina) [inline lib/driverProviders/browserStack.ts]: Done.

File: lib/exitCodes.ts (hunk 6 above)
Comment (cnishina) [inline lib/exitCodes.ts]: Removing the default message and plan to use the message from the thrown error.

File: lib/exitCodes.ts (hunk 5 above)
Comment (cnishina) [inline lib/exitCodes.ts]: This was a weird work around because `stack` is not being part of `node` typings. Using `captureStackTrace` with `any` to get the stack.

File: lib/exitCodes.ts (hunk 4 above)
Comment (cnishina) [inline lib/exitCodes.ts]: Done

File: lib/exitCodes.ts (hunk 7 above)
Comment (cnishina) [inline lib/exitCodes.ts]: Removed.

File: lib/launcher.ts (hunk 8 above)
Comment (cnishina) [inline lib/launcher.ts]: Removed.

File: lib/ptor.ts [hunk 9]
```diff
@@ -2,7 +2,11 @@ import {Browser, ElementHelper} from './browser';
 import {ProtractorBy} from './locators';
```
Comment (juliemr) [inline lib/ptor.ts]: This looks like an unrelated change.

Comment (cnishina) [inline lib/ptor.ts]: Yes, I need to stop sneaking these in. Will revert this file back.

**Comment by juliemr** [comment]
LGTM once that's done.
